crates/core/src/error/collector.rs

```diff
@@ -0,0 +1,289 @@
+use derive_more::From;
+
+use crate::{error::RusticErrorKind, RusticError};
+use log::{error, info, warn};
+
+/// A rustic issue result
+///
+/// rustic issue results are used to return a result along with possible issues.
+#[derive(Debug)]
+pub enum RusticIssueResult<T> {
+    /// For when we return a result along with possible issues
+    Ok(T, Option<Vec<RusticIssue>>),
+
+    /// For when we abort and return only the errors that occurred
+    Err(Vec<RusticError>),
+}
+
+/// A rustic issue
+///
+/// An issue is a message that can be logged to the user.
+#[derive(Debug)]
+pub enum RusticIssue {
```

Comment:
Why not using something like `error: RusticError` and `severity: RusticSeverity` with the latter being an enum? I can imagine we can have the exactly identical error which is at some place classified as warning (or even info) and at another place just an error.

Reply:
I don't know, I would not classify an error at one place as a warning and at the other as an info. An error is an error, and should be clearly communicated as one. If we change the general meaning of an error, then it shouldn't be an error in the first place.

Reply:
But I do know what you mean, something like: 'File not found'. At one place, it can be irrecoverable, e.g. a config file not found for a backend. At some other place, it maybe can be continued with default settings.

But in that case we should handle that error in-place and not return an error I feel.

Reply:
Also, important to say here: `RusticIssue`, `RusticWarning`, and `RusticIssueCollector` should never go over the library boundary. So it's for internal use only, it should not be part of our public API, and we should actually not return `RusticIssue` from any public function. It should only be used for internal error handling, so we as library authors can judge situations better internally, and give our users errors only in cases that are worthwhile.

Hence, I reduced the visibility of all the types in this file to `pub(crate)`